### ISSUE TITLE
JsonDecoder does not correctly escape four-byte UTF-8 characters

### ISSUE THREAD

**Issue by lluchs** [xp-framework/webservices#1]
For example, the :hankey: emoji is U+1f4a9, but is escaped as \uf4a9. JSON does not properly support four-byte escapes and needs two of them in this case.

**Comment by thekid** [comment]
```sh
$ python -c 'import json; str = u"Poop = \U0001f4a9"; print "String: `%s`\nJSON: %s\n" % (str, json.dumps(str));'
String: `Poop = 💩`
JSON: "Poop = \ud83d\udca9"
```

Right?

**Comment by lluchs** [comment]
Yes, this is what I'd expect. The code [here](https://github.com/xp-framework/webservices/blob/master/src/main/php/webservices/json/JsonDecoder.class.php#L101) does not seem to handle this case though.

**Comment by thekid** [comment]
OK, what's missing is handling of UTF-16 surrogate pairs. Quoting [Wikipedia](https://en.wikipedia.org/wiki/UTF-16#U.2B10000_to_U.2B10FFFF)

> Code points from the other planes (called Supplementary Planes) are encoded as two 16-bit code units called surrogate pairs [...]

Also nicely explained in http://www.2ality.com/2013/09/javascript-unicode.html under *Refering to astral plane characters via escapes.*

```javascript
var str = '🐄';
str === '\uD83D\uDC04'  // true
```

Calculating the character code is done by combining the head and the tail of the surrogate pair as per  http://unicode.org/faq/utf_bom.html#utf16-4

**Comment by thekid** [comment]
The commits above are against the new JSON library; this library still has its own implementation which would need separate fixing.

**Comment by thekid** [comment]
Released fix for this in `xp-forge/json`, version [2.0.1](https://github.com/xp-forge/json/releases/tag/v2.0.1)

**Comment by thekid** [comment]
Fixed in `xp-framework/rest`, version [7.2.0](https://github.com/xp-framework/rest/releases/tag/v7.2.0)
